File: leetcode/tree/545_Boundary_of_Binary_Tree/codes/tree.c
/**
Definition for a binary tree node.
 */

struct TreeNode {
    int val;
    struct TreeNode *left;
    struct TreeNode *right;
};

bool isLeaf(struct TreeNode* node) {
    return node && !node -> left && !node -> right;
}

void addLeftBoundary(struct TreeNode* node, int* res, int* size) {
    while (node) {
        if (!isLeaf(node)) res[(*size)++] = node -> val;
        node = node -> left ? node -> left : node -> right;
    }
}

void addRightBoundary(struct TreeNode* node, int* stack, int* top) {
    while (node) {
        if (!isLeaf(node)) stack[(*top)++] = node -> val;
        node = node -> right ? node -> right : node -> left;
    }
}

void addLeaves(struct TreeNode* node, int* res, int* size) {
    if (!node) return;
    if (isLeaf(node)) {
        res[(*size)++] = node -> val;
        return;
    }
    addLeaves(node -> left, res, size);
    addLeaves(node -> right, res, size);
}

/**
Note: The returned array must be malloced, assume caller calls free().
 */
int* boundaryOfBinaryTree(struct TreeNode* root, int* returnSize){
    int* res = malloc(10000sizeof(int));
    int* rightStack = malloc(10000sizeof(int));
    int size = 0, top = 0;

    if (!root) {
        *returnSize = 0;
        return res;
    }

    if (!isLeaf(root)) res[size++] = root -> val;
    addLeftBoundary(root -> left, res, &size);
    addLeaves(root, res, &size);
    addRightBoundary(root -> right, rightStack, &top);

    for (int i = top - 1; i >= 0; i--)
        res[size++] = rightStack[i];

    *returnSize = size;
    free(rightStack);
    return res;
}

File: leetcode/tree/545_Boundary_of_Binary_Tree/codes/tree.py
from typing import List, Optional

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def boundaryOfBinaryTree(self, root: Optional[TreeNode]) -> List[int]:
        if not root:
            return []

        def isLeaf(node):
            return node and not node.left and not node.right

        def addLeftBoundary(node):
            while node:
                if not isLeaf(node):
                    boundary.append(node.val)
                node = node.left if node.left else node.right

        def addRightBoundary(node):
            stack = []
            while node:
                if not isLeaf(node):
                    stack.append(node.val)
                node = node.right if node.right else node.left
            boundary.extend(reversed(stack))

        def addLeaves(node):
            if not node:
                return
            if isLeaf(node):
                boundary.append(node.val)
                return
            addLeaves(node.left)
            addLeaves(node.right)

        boundary = [root.val] if not isLeaf(root) else []
        addLeftBoundary(root.left)
        addLeaves(root)
        addRightBoundary(root.right)
        return boundary
